pkg/agent/multicast/multicast_route.go

```diff
@@ -0,0 +1,423 @@
+	if err != nil {
+		return err
+	}
+	routeEntry := &inboundMulticastRouteEntry{
+		group: group.String(),
+		src:   src.String(),
+		vif:   inboundVif,
+	}
+	c.inboundRouteCache.Add(routeEntry)
+	return nil
+}
+
+// inboundMulticastRouteEntry encodes the inbound multicast routing entry.
+// e.g. (10.0.0.55,226.94.9.9)           Iif: wlan0      Oifs: antrea-gw0
+// the oif is alway gatewayInterface so we do not put it in the struct
+type inboundMulticastRouteEntry struct {
+	group string
+	src   string
+	vif   uint16
+}
+
+// outboundMulticastRouteEntry encodes the outbound multicast routing entry.
+// e.g. (10.244.1.4,224.1.2.10)          Iif: antrea-gw0 Oifs: wlo1 wlp3s0 zt5u4zjljr
+// the oif is alway gatewayInterface so we do not put it in the struct
+type outboundMulticastRouteEntry struct {
+	group string
+	src   string
+	vifs  []uint16
+}
+
+func getMulticastInboundEntryKey(obj interface{}) (string, error) {
+	entry := obj.(*inboundMulticastRouteEntry)
+	return entry.group + "/" + entry.src + "/" + string(entry.vif), nil
+}
+
+func getMulticastOutboundEntryKey(obj interface{}) (string, error) {
+	entry := obj.(*outboundMulticastRouteEntry)
+	return entry.group + "/" + entry.src, nil
+}
+
+func inboundvifIndexFunc(obj interface{}) ([]string, error) {
+	entry, ok := obj.(*inboundMulticastRouteEntry)
+	if !ok {
+		return []string{}, nil
+	}
+	return []string{strconv.FormatUint(uint64(entry.vif), 10)}, nil
+}
+
+func oudboundvifsIndexFunc(obj interface{}) ([]string, error) {
+	entry, ok := obj.(*outboundMulticastRouteEntry)
+	if !ok {
+		return []string{}, nil
+	}
+	vifs := make([]string, len(entry.vifs))
+	for _, vif := range entry.vifs {
+		vifs = append(vifs, strconv.FormatUint(uint64(vif), 10))
+	}
+	return vifs, nil
+}
+
+func inboundGroupIndexFunc(obj interface{}) ([]string, error) {
+	entry, ok := obj.(*inboundMulticastRouteEntry)
+	if !ok {
+		return []string{}, nil
+	}
+	return []string{entry.group}, nil
+}
+
+func outboundGroupIndexFunc(obj interface{}) ([]string, error) {
+	entry, ok := obj.(*outboundMulticastRouteEntry)
+	if !ok {
+		return []string{}, nil
+	}
+	return []string{entry.group}, nil
+}
+
+// getExternalMulticastIPs tries to get external multicast interfaces for configuring outbound multicast routes
+// Currently the implementation is to filter out interfaces from ifaceStore and mulicast disbaled interfacs
+func (c *MRouteClient) getExternalMulticastIPs() (map[string][]net.IP, error) {
+	interfacesMap := make(map[string]bool)
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return nil, err
+	}
+	for _, iface := range c.ifaceStore.GetAllInterfaces() {
+		interfacesMap[iface.InterfaceName] = true
+	}
+	interfacesMap[c.gatewayInterface] = true
+	addrMap := make(map[string][]net.IP)
+
+	for _, iface := range ifaces {
+		_, ok := interfacesMap[iface.Name]
+		// the interface should be multicast enabled
+		// and not in the ifaceStore
+		if ok || !strings.Contains(iface.Flags.String(), "multicast") {
```

Comment:
Do we have a more efficient way to check this? At least define a constant for "multicast".

Reply:
After offline discussion with @wenyingd , we agreed allowing users manually set external multicast interfaces in the antrea config file would be a better and simpler solution. If set in the antrea config file, this method could be removed and no need to detect and handle the multicast interfaces change. what's your opinion?  Thanks

Reply:
That sounds good to me. But considering in many cases the Node just has one interface, do we have a way to look up and use that interface by default?

Reply:
I think you the `one interface` you mentioned  is transportInterface. I have added `externalMulticastInterface` in the antrea agent config file. Basically the result of `externalMulticastInterface union TransportInterface` will be the multicast interfaces we will use to route multicast traffic.

Reply:
No, TransportInterface is also optional. When the Node has only one interface, TransportInterface parameter is useless, and so will not be set either.

BTW, when externalMulticastInterface is set, we should just use it; only when it is not set we consider TransportInterface next.

Reply:
Updated.

 Basically, we need one interface to route multicast traffic to remote Nodes base on the  order of `transportInterface`, `transportInterfaceCIDRs` and `The Node IP`,  which is `NodeTransportName` calculated in `NodeConfig`. Also user can specify `externalMulticastInterfaces` to route multicast traffic to the outside world. the result of `externalMulticastInterface union [NodeTransportName]` will be all the multicast interfaces to route multicast traffic.

Please correct me If I understand it incorrectly, thanks.